### PR TITLE
[DMS-1264] [DMS-1266]: HTTP If-None-Match support (conditional GET 304 + write create-guard)

### PR DESCRIPTION
**Pull request by stephenfuqua**
## Summary

Implements HTTP `If-None-Match` support ([DMS-1264](https://edfi.atlassian.net/browse/DMS-1264)), building on the etag→`ContentVersion` work (DMS-1252). Two parts:

- **Conditional GET (read path).** GET-by-id now emits an `ETag` response header and returns `304 Not Modified` when the client's `If-None-Match` matches the current representation. Comparison is against the **full served tag** (representation-sensitive: a different profile / format / links mode does not match — deliberately distinct from the write-side projection). Weak (`W/`) and unquoted values are accepted; a bare `*` returns `304` when the resource exists.
- **Write create-guard (write path).** `If-None-Match: *` is insert-only — a POST resolving to an insert proceeds; a POST resolving to an existing document, or a PUT against an existing resource, returns `412`; a PUT to a missing target returns `404` (not `412`). A specific tag returns `412` when it matches the current representation and proceeds otherwise (state-significant projection). `If-Match` takes precedence when both headers are present. Honored on both the relational and descriptor paths, on both the before-authorization and deferred-post-authorization routes; `DELETE` ignores `If-None-Match`.

In addition, this implements `ETag` response headers on GET-by-ID requests ([DMS-1266](https://edfi.atlassian.net/browse/DMS-1266)).

Behavior is specified in the ADR amendment "Amendment (2026-07-06): If-None-Match support" (`reference/adr-etag-from-content-version.md`) and reflects the **ODS-6853** legacy finding: legacy ODS/API returns `304` only for an *unquoted* `If-None-Match`, so unquoted acceptance is a compatibility requirement (the server still emits quoted strong ETags).

## Base branch

Targets **`etag-content-version`**, not `main`, so reviewers can merge the base etag work with or without this `If-None-Match` feature.

## Test plan

- [x] Backend unit: **1920** pass
- [x] Core unit: **2582** pass · Frontend unit: **218** pass
- [x] Cross-engine integration (PostgreSQL **and** SQL Server): conditional-GET `304` (10) + write create-guard (14), including the **deferred-authorization** write path on both engines
- [x] E2E scenarios added (build-verified locally; executed in CI)

## Out of scope

`If-Modified-Since`, `If-Unmodified-Since`, and `If-Range` remain unimplemented (date-based validators are deferred in favor of the strong `ContentVersion`-derived entity tag; `If-Range` needs HTTP range support the DMS does not provide).

## AI-use disclosure

This feature was implemented and reviewed with substantial AI assistance (Claude Code). Each task passed a two-stage (spec-compliance + code-quality) review, plus a whole-feature final review for each of the read and write parts. Human review is required before merge; accountability for the final result rests with the development team.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[DMS-1264]: https://edfi.atlassian.net/browse/DMS-1264?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DMS-1266]: https://edfi.atlassian.net/browse/DMS-1266?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by bradbanister** [comment]
### Review

**Findings**

1. **Medium: `If-None-Match` does not support entity-tag lists.**
   The ADR says “full If-None-Match support,” but [EtagValue.cs](/home/brad/work/dms-root/DMS-1264/src/dms/core/EdFi.DataManagementService.Core/Utilities/EtagValue.cs:51) explicitly parses only one value. [GetByIdHandler.cs](/home/brad/work/dms-root/DMS-1264/src/dms/core/EdFi.DataManagementService.Core/Handler/GetByIdHandler.cs:157) and [WritePreconditionFactory.cs](/home/brad/work/dms-root/DMS-1264/src/dms/core/EdFi.DataManagementService.Core/Backend/WritePreconditionFactory.cs:40) treat the whole header as a single tag. A compliant request like `If-None-Match: "stale", "current"` would miss on GET and return `200`, or allow a write, instead of matching any listed tag and returning `304`/`412`. [AspNetCoreFrontend.cs](/home/brad/work/dms-root/DMS-1264/src/dms/frontend/EdFi.DataManagementService.Frontend.AspNetCore/AspNetCoreFrontend.cs:74) also drops repeated physical header values by taking only the first nonblank value.

2. **Low: `If-None-Match` 412 responses still say `If-Match`.**
   If an `If-None-Match` write fails, the outward problem text still says the mismatch was against the `If-Match` header. See [UpsertHandler.cs](/home/brad/work/dms-root/DMS-1264/src/dms/core/EdFi.DataManagementService.Core/Handler/UpsertHandler.cs:140) and [UpdateByIdHandler.cs](/home/brad/work/dms-root/DMS-1264/src/dms/core/EdFi.DataManagementService.Core/Handler/UpdateByIdHandler.cs:83). The status code is correct, but the response body is inaccurate for create-guard failures like `If-None-Match: *` on an existing resource.

3. **Low: `EtagPreconditionEvaluator` fails open for future precondition arms.**
   [EtagPreconditionEvaluator.cs](/home/brad/work/dms-root/DMS-1264/src/dms/backend/EdFi.DataManagementService.Backend/Etag/EtagPreconditionEvaluator.cs:18) returns `true` for anything other than `IfMatch`/`IfNoneMatch` via `_ => true`. Since this is central security/concurrency-adjacent write precondition logic, prefer explicit `WritePrecondition.None => true` and throw for unknown arms.

**Simplicity Notes**

- `TryParseConditionalTag` returns `false` only for null/empty input, so fallback branches in callers are effectively dead for real ASP.NET requests. This would be cleaner as a normalizer that returns a parsed value, or as part of a real list parser.

- `IfMatchPreconditionEvaluation` / `DeferredIfMatch` now apply to both `If-Match` and `If-None-Match`; renaming them to ETag/precondition-neutral names would reduce mental overhead.

- The new Get-by-id unit tests repeat the same nested repository/setup/assertion structure across many `If-None-Match` cases. Parameterized cases over header value, expected status, and body/etag behavior would be simpler and make list-case coverage cheaper.

- The new E2E `If-None-Match` header helpers duplicate existing generic “with header name/value” steps. Low risk, but it adds another path to keep in sync with auth/content-type behavior.

**Open Question**

Jira says `If-Match` takes precedence when both headers are present. The branch covers that for writes, but GET only checks `If-None-Match`. The ADR says there is no read-side `If-Match` today, so this may be intentional; worth confirming.

**Comment by bradbanister** [comment]
If it's intentional to have no entity-tag list support, then the ticket/design needs updating.

**Comment by stephenfuqua** [comment]
> If it's intentional to have no entity-tag list support, then the ticket/design needs updating.

Although the ODS/API does _not_ support the entity-tag list, this is a good opportunity for DMS to get it right. Good call out.

File: reference/adr-etag-from-content-version.md [hunk 1]
```diff
@@ -3,12 +3,15 @@
 **Status:** Accepted — implemented in the relational backend; the three design docs
 (`update-tracking.md`, `transactions-and-concurrency.md`, `flattening-reconstitution.md`) have been
 updated to match. \
+**Date:** 2026-06-30 (accepted 2026-07-03) \
 **Amended 2026-07-04:** `profileCode` removed from the `If-Match` comparison to restore legacy
 compatibility — see [Amendment (2026-07-04)](#amendment-2026-07-04-profilecode-removed-from-if-match-comparison). \
 **Amended 2026-07-05:** unquoted `If-Match` values accepted as equivalent to quoted ones for legacy
 compatibility — see [Amendment (2026-07-05)](#amendment-2026-07-05-unquoted-if-match-values-accepted-as-equivalent-to-quoted). \
-**Amended 2026-07-05:** bare `If-Match: *` honored as an existence precondition per RFC 7232 §3.1 —
+**Amended 2026-07-05:** bare `If-Match: *` honored as an existence precondition per RFC 9110 §3.1 —
```
Comment (stephenfuqua) [inline reference/adr-etag-from-content-version.md]: 7232 is still widely cited, but it was superseded by 9110.

**Comment by stephenfuqua** [comment]
# PR #1095 Review — HTTP `If-None-Match` support (conditional GET 304 + write create-guard)
> DMS-1264 (Story) + DMS-1266 (Bug): honor `If-None-Match` per RFC 9110, fix missing `ETag` header on GET-by-id
> Base branch: `etag-content-version` (not `main`)
> Reviewed by: Security · Functionality · Maintainability · Usability · Test Coverage

> [!TIP]
> Stephen added personal analysis with these signals:
> ✅ agree - Stephen or someone else should resolve
> ⚠️ maybe - Stephen is fine with ignoring the finding, but doesn't object to resolving it
> ❌ disagree / won't do - Stephen thinks we should not do anything

---

## Critical
None found.

## High
- **[Maintainability, Functionality, Usability]** 412 responses caused by an `If-None-Match` precondition failure return an error body that blames the wrong header and describes the wrong problem. `FailureResponse.ForETagMisMatch` has only two reasons (`Concurrency`, `TargetDoesNotExist`); every `If-None-Match` create-guard failure defaults to `Concurrency`, producing `"detail": "The item has been modified by another user."` and an error citing the `'If-Match'` header — even though the client sent `If-None-Match` and nothing was concurrently modified. No test asserts 412 body text for the new `If-None-Match` scenarios (E2E scenarios 16/17 check status code only).
  Fix: Add a distinct failure reason (e.g. `AlreadyExists`) produced when the triggering precondition is `IfNoneMatch`, and give `FailureResponse.ForETagMisMatch` an `If-None-Match`-specific message. Add body-text assertions to the corresponding E2E scenarios. ✅ SF: We should fix this.

## Medium
- **[Usability]** The current server-side ETag is computed (`RelationalCurrentEtagPreconditionCheckResult.CurrentEtag`) specifically to support this feature, but is only used for logging/tests — it's never returned as a response header on 412, forcing clients into an extra GET before they can retry a conflicting write. ❌ SF: WON'T DO. This is normal and correct behavior.
- **[Test Coverage]** Descriptor resources are never exercised against a live PostgreSQL/SQL Server engine for If-None-Match — only the relational (Student) resource has DB-backed integration scenarios; descriptor coverage is mock-only. ✅ SF: We should fix this.
- **[Test Coverage, Functionality, Usability]** An RFC 9110-invalid list (`If-None-Match: *, "tag"` — wildcard mixed with other tags) is silently treated as a literal, always-non-matching tag rather than rejected or explicitly handled. Behavior is safe (fails closed to "no match") but undocumented in the ADR and untested at every layer. ⚠️ SF: This is such an edge that I don't think it is worthwhile to hard-code this behavior into tests.
- **[Test Coverage]** The "header present but empty" boundary case has explicit code branches in both `GetByIdHandler` and `WritePreconditionFactory` but no end-to-end test exercising it. ⚠️ SF: I can live with that, if we have unit tests.
- **[Maintainability]** Debug logging around the (correctly) centralized precondition decision is inconsistent: the descriptor and DELETE paths log the sanitized client tag, but the highest-traffic relational PUT/POST path does not. ✅ SF: We should add this debug logging for consistency.
- **[Usability]** The full `If-None-Match` contract (list support, weak/unquoted tolerance, inverted wildcard semantics vs. `If-Match`, the 404-vs-412 split on missing targets) is documented only in the internal ADR (`reference/`), not surfaced anywhere a client-facing API consumer would read it. ❌ SF: To be expected, since our client-facing documentation is in another repository.

## Low
- **[Security]** No application-level cap on entity-tag list size/length in `EtagValue.ParseConditionalTagList` — relies entirely on Kestrel's default header-size limit rather than a feature-aware bound. ⚠️ SF: Good enough for me.
- **[Security]** Comma-splitting for entity-tag lists doesn't honor RFC 9110 quoted-string escaping; a comma inside a quoted tag would mis-split (fails safe — degrades to non-match — but not spec-correct). ⚠️ SF: truly low value, given that our etag format does not use commas. I don't think this is worth addressing. But I won't object if someone defensively protects against this situation.
- **[Maintainability]** Bare `*` wildcard raw-header detection (`rawValue == "*"`) is independently reimplemented three times across `WritePreconditionFactory` and `GetByIdHandler` rather than shared via one helper. ✅ SF: would be nice to address this.
- **[Maintainability]** Naming collision: `EtagPreconditionEvaluator` (does the check) vs. `EtagPreconditionEvaluation` (enum for *when* to check) — easy to confuse, used together in the same methods. ✅ SF: would be nice to address this.
- **[Maintainability]** Asymmetric abstraction — write-side precondition logic was extracted into `EtagPreconditionEvaluator`; the read-side equivalent stays inlined in `GetByIdHandler` with only a comment cross-referencing the split. ✅ SF: would be nice to address this.
- **[Usability]** `DELETE` silently ignores `If-None-Match` (by design, per ADR) with no client-visible signal that the header had no effect. ⚠️ SF: I can live with the current behavior, but also wouldn't object to someone addressing it.
- **[Usability, Test Coverage]** Malformed/garbage `If-None-Match` values (whitespace-only, unmatched quotes) never produce a 4xx — they silently degrade to "no precondition" — and this path lacks explicit test coverage. ⚠️ SF: I can live with the current behavior, but also wouldn't object to someone addressing it.
- **[Usability]** The "two clients race to claim a create via `If-None-Match: *`" scenario isn't covered by any test, so the loser's actual response code is unverified (likely resolves through pre-existing conflict machinery, but unconfirmed). ✅ SF: Should think about this a little more.

## Nit
- **[Security]** Served `_etag` embeds `ContentVersion` + `schemaEpoch` — informational only, neither is sensitive.  ❌ SF: Yes, by design, no action to take
- **[Maintainability]** Mixed parameter casing (`value`, `IsWildcard`) in `WritePrecondition.IfNoneMatch`'s convenience constructor. ❌ SF: this is either a hallucination, or it was changed in a refactor for another finding.
- **[Functionality]** `etag.feature` scenario numbering skips "07" — cosmetic only. ✅ SF: should fix.
- **[Test Coverage]** Some near-duplicate test bodies in `DescriptorWriteHandlerPreconditionTests.cs` could be parameterized to reduce the ~2500 lines of test additions. ⚠️ SF: no personal preference.

---

## Specialist Verdicts
| Reviewer | Verdict |
|----------|---------|
| Security | No Critical/High/Medium issues — authorization ordering, error hygiene, and log sanitization are all correctly and defensively implemented. |
| Functionality | PARTIAL — every acceptance criterion in DMS-1264/DMS-1266 is correctly implemented, but the misleading 412 error message should be fixed before merge. |
| Maintainability | The core precondition-decision refactor is a solid, verified consolidation with no dead code, but surrounding logging/messaging layers didn't keep pace with the widened If-Match/If-None-Match model. |
| Usability | RFC 9110 semantics are implemented carefully and well-tested, but the error surface (wrong header blamed, discarded current-etag, ADR-only docs) falls short of "actionable" for API consumers. |
| Test Coverage | ADEQUATE — happy paths, list/wildcard semantics, cross-engine parity, and both regression fixes are thoroughly locked in; three gaps (spec-invalid wildcard-list, empty-header boundary, descriptor-vs-real-DB) keep it short of STRONG. |

---

## Summary
The RFC 9110 semantics themselves — conditional GET 304/200, the write create-guard's 412/404 split, wildcard inversion, If-Match precedence, weak/unquoted/list parsing, cross-engine parity — are implemented correctly and thoroughly tested, with no security concerns. The one thing worth fixing before merge is the 412 error message bug (three reviewers flagged it independently): it currently blames "If-Match" and describes a concurrency conflict for what is actually an `If-None-Match` create-guard failure, which is directly misleading for the exact feature this PR ships. Also worth addressing soon (if not blocking): returning the current ETag on 412 so clients can retry without an extra GET, and adding descriptor-resource integration coverage against real PostgreSQL/SQL Server.

---
*This review was generated with substantial AI assistance (Claude Code, `team-pr-review` skill — five parallel specialist passes: security, functionality, maintainability, usability, test coverage). Human review and judgment is still required before acting on these findings.*

File: src/dms/core/EdFi.DataManagementService.Core/Handler/GetByIdHandler.cs [hunk 2]
```diff
@@ -120,7 +135,64 @@ private static FrontendResponse CreateSuccessResponse(RequestInfo requestInfo, J
...
+        // since reaching this method means the resource exists, the precondition is false. This must be
+        // detected from the RAW header value (before parsing), because EtagValue.TryParseConditionalTag
+        // strips quotes and would otherwise turn a quoted "*" (an ordinary opaque tag) into the wildcard.
+        bool isWildcard = string.Equals(rawHeaderValue, "*", StringComparison.Ordinal);
+
+        IReadOnlyList<string> clientTags = isWildcard
+            ? []
+            : EtagValue.ParseConditionalTagList(rawHeaderValue);
+        if (!isWildcard && clientTags.Count == 0)
+        {
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply src/dms/core/EdFi.DataManagementService.Core/Handler/GetByIdHandler.cs:195]: This suggestion incorrectly deletes `{` on line 174. I'll apply this and related changes by hand.

File: src/dms/core/EdFi.DataManagementService.Core/Backend/WritePreconditionFactory.cs [hunk 3]
```diff
@@ -11,29 +11,47 @@ namespace EdFi.DataManagementService.Core.Backend;
...
+            }
+
+            // Normalize the wire form to the opaque tag the backend compares against: strip the
+            // surrounding quotes of a strong entity-tag and tolerate a bare unquoted value. A weak (W/)
+            // validator is rejected by the helper and kept verbatim so the backend's state-significant
+            // projection cannot equal a well-formed current tag (RFC 7232 §3.1: weak validators must not
+            // be used with If-Match).
+            return EtagValue.TryParseHeaderValue(ifMatchValue, out var opaqueTag)
+                ? new WritePrecondition.IfMatch(opaqueTag)
+                : new WritePrecondition.IfMatch(ifMatchValue ?? string.Empty);
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply src/dms/core/EdFi.DataManagementService.Core/Backend/WritePreconditionFactory.cs]: Will address this locally.

File: src/dms/core/EdFi.DataManagementService.Core/Backend/WritePreconditionFactory.cs [hunk 4]
```diff
@@ -11,29 +11,47 @@ namespace EdFi.DataManagementService.Core.Backend;
...
+            if (string.Equals(ifNoneMatchValue, "*", StringComparison.Ordinal))
+            {
+                return new WritePrecondition.IfNoneMatch("*", IsWildcard: true);
+            }
+
+            // Weak comparison: accept W/ (unlike If-Match) and tolerate unquoted for each list element.
+            var tags = EtagValue.ParseConditionalTagList(ifNoneMatchValue);
+            return tags.Count > 0
+                ? new WritePrecondition.IfNoneMatch(tags)
+                : new WritePrecondition.IfNoneMatch(ifNoneMatchValue ?? string.Empty);
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply src/dms/core/EdFi.DataManagementService.Core/Backend/WritePreconditionFactory.cs:52]: Will address this locally.

File: reference/adr-etag-from-content-version.md [hunk 5]
```diff
@@ -3,12 +3,15 @@
...
 see [Amendment (2026-07-05, wildcard)](#amendment-2026-07-05-if-match-wildcard-matching). \
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply reference/adr-etag-from-content-version.md]: Will address this locally.

**Review by simpat-adam — APPROVED** [review]
